src/categorical_algebra/StructAcsets.jl

```diff
@@ -0,0 +1,356 @@
+module StructAcsets
+export @acset_type, add_parts!, set_subpart!, schema, Arrow, ParsedSchema
+
+using MLStyle
+
+using ...Theories, ...Present, ...Syntax
+using ...Meta: strip_lines
+
+# Interface for StructAcsets
+############################
+
+abstract type AbstractStructAcset end
+
+""" Primitive operation for adding parts (without subparts).
+
+Override this function in generated code.
+"""
+function _add_parts!() end
+
+""" A non-generated convenience wrapper around _add_parts!
+"""
+add_parts!(acs::AbstractStructAcset, ob::Symbol, n::Int) = _add_parts!(acs, Val{ob}, n)
+
+""" Primitive operation for setting subparts, including setting indexes
+
+Override this function in generated code
+"""
+function _set_subpart!() end
+
+""" A non-generated convenience wrapper around _set_subpart!
+"""
+set_subpart!(acs::AbstractStructAcset, part::Int, f::Symbol, subpart) =
+  _set_subpart!(acs, part, Val{f}, subpart)
+
+function set_subpart!(acs::AbstractStructAcset, parts::Vector{Int}, f::Symbol, subparts::Vector)
+  for (x,y) in zip(parts,subparts)
+    set_subpart!(acs, x, f, y)
+  end
+end
+
+""" Returns the number of parts for an object. Doesn't need to be generated.
+"""
+nparts(acs::AbstractStructAcset, ob::Symbol) = acs.obs[ob][]
+
+""" Override this function in generated code
+"""
+function _apply_acset() end
+
+""" This is "functor syntax" for acsets.
+"""
+(acs::AbstractStructAcset)(x::Symbol) = _apply_acset(acs, Val{x})
+
+# More Convenient/Efficient Specialized Schema Representation
+#############################################################
+
+""" Can be used to represent Hom or Attr
+"""
+struct Arrow
+  name::Symbol
+  dom::Symbol
+  codom::Symbol
+  idxed::Bool
+end
+
+""" Same information as Presentation{Schema}, but without equations,
+and in a much more concrete/easily accessible format.
+"""
+struct ParsedSchema
```

Comment:
I guess we can't use an ACSet for this... too bad the CT primordial soup can't be implemented as a soup.